src/consensus/cfund.cpp

```diff
@@ -501,22 +574,97 @@ std::string CFund::CProposal::GetState(uint32_t currentTime) const {
     return sFlags;
 }
 
+flags CFund::CProposal::GetLastState() const {
+    flags ret = NIL;
+    int nHeight = 0;
+    for (auto& it: mapState)
+    {
+        if (mapBlockIndex.count(it.first) == 0)
+            continue;
+
+        if (mapBlockIndex[it.first]->nHeight > nHeight)
+        {
+            nHeight = mapBlockIndex[it.first]->nHeight;
+            ret = it.second;
+        }
+    }
+    return ret;
+}
+
+CBlockIndex* CFund::CProposal::GetLastStateBlockIndex() const {
+    CBlockIndex* ret = nullptr;
+    int nHeight = 0;
+    for (auto& it: mapState)
+    {
+        if (mapBlockIndex.count(it.first) == 0)
+            continue;
+
+        if (!chainActive.Contains(mapBlockIndex[it.first]))
+            continue;
+
+        if (mapBlockIndex[it.first]->nHeight > nHeight)
+        {
+            nHeight = mapBlockIndex[it.first]->nHeight;
+            ret = mapBlockIndex[it.first];
+        }
+    }
+    return ret;
+}
+
+CBlockIndex* CFund::CProposal::GetLastStateBlockIndexForState(flags state) const {
+    CBlockIndex* ret = nullptr;
+    int nHeight = 0;
+    for (auto& it: mapState)
+    {
+        if (it.second != state)
+            continue;
+
+        if (mapBlockIndex.count(it.first) == 0)
+            continue;
+
+        if (!chainActive.Contains(mapBlockIndex[it.first]))
+            continue;
+
+        if (mapBlockIndex[it.first]->nHeight > nHeight)
+        {
+            nHeight = mapBlockIndex[it.first]->nHeight;
+            ret = mapBlockIndex[it.first];
+        }
+    }
+    return ret;
+}
+
+bool CFund::CProposal::SetState(const CBlockIndex* pindex, flags state) {
+    mapState[pindex->GetBlockHash()] = state;
+    return true;
+}
+
+bool CFund::CProposal::ClearState(const CBlockIndex* pindex) {
+    mapState.erase(pindex->GetBlockHash());
+    return true;
+}
+
 void CFund::CProposal::ToJson(UniValue& ret, CCoinsViewCache& coins) const {
     AssertLockHeld(cs_main);
 
+    flags fState = GetLastState();
+    uint256 blockhash;
+    CBlockIndex* pblockindex = GetLastStateBlockIndex();
+    if (pblockindex) blockhash = pblockindex->GetBlockHash();
+
     ret.pushKV("version", nVersion);
     ret.pushKV("hash", hash.ToString());
     ret.pushKV("blockHash", txblockhash.ToString());
     ret.pushKV("description", strDZeel);
     ret.pushKV("requestedAmount", FormatMoney(nAmount));
     ret.pushKV("notPaidYet", FormatMoney(GetAvailable(coins)));
+    ret.pushKV("notRequestedYet", FormatMoney(GetAvailable(coins, true)));
     ret.pushKV("userPaidFee", FormatMoney(nFee));
     ret.pushKV("paymentAddress", Address);
     if(nVersion >= 2) {
         ret.pushKV("proposalDuration", (uint64_t)nDeadline);
-        if (fState == ACCEPTED && mapBlockIndex.count(blockhash) > 0) {
-            CBlockIndex* pBlockIndex = mapBlockIndex[blockhash];
-            ret.pushKV("expiresOn", pBlockIndex->GetBlockTime() + (uint64_t)nDeadline);
+        if ((fState == ACCEPTED || fState == PAID) && pblockindex) {
```

Comment:
won't `pblockindex` always evaluate true unless the proposal has not got any state at all? Is that what you're supposed to be checking in this condition? Or should we be using `GetLastStateBlockIndexForState()` to check a particular state has been correctly achieved?

Reply:
it's true that `(fState == ACCEPTED || fState == PAID)` can't give us a null pointer in `pblockindex`, but i think the check does not hurt for sanity